### DISCUSSION TITLE
Cooperative scheduling vs Preemptive scheduling

### DISCUSSION THREAD

**Discussion by 04116** [04116/til#44] [General]
## lý thuyết
> nói về cách runtime phân bổ thời gian sử dụng CPU (thứ có giới hạn) cho các process/thread/goroutine/task,.. (rất nhiều)
- Cooperative scheduling: process được làm xong việc của mình
- Preemptive scheduling: process được allow **FIXED time/quota**, hết quota; dù có chưa xong việc cũng sẽ bị pause lại, nhường cho cho process khác

## ứng dụng
- ~~golang dùng Cooperative~~; erlang/OTP dùng Preemptive
    - ~~1 CPU bounded task ở golang sẽ được tạo điều kiện để chạy... cho đến khi nó muốn pause~~
        - ~~trên thực tế, kernel bắt buộc các thread sẽ chạy ở Preemptive; và golang runtime sẽ xử lý để khi phía OS switch context; phía goroutine không ảnh hưởng~~
    - ở erlang/OTP các process sẽ bị pause lại khi hết quota
> - ~~golang tốt hơn cho các task cần CPU bounded; nhưng khi 1 task quá bounded :v thì có thể ảnh hưởng đến goroutine khác => ảnh hưởng đến latency của hệ thống~~
> - erlang/OTP rất tốt cho ứng dụng cần latency ổn định!!! (switch, voice,...)

Update: golang đã đang chạy ở Preemptive
cred @harry-2401 

## micro-bench
[0001-Initial-benchmark-commit.patch](https://github.com/user-attachments/files/21078611/0001-Initial-benchmark-commit.patch)
[0001-Add-Makefile-for-benchmarks.patch](https://github.com/user-attachments/files/21078612/0001-Add-Makefile-for-benchmarks.patch)

```javac Bench.java
Running Java benchmark...
java Bench
Loop ops/sec: 4478787708
SHA-256 ops/sec: 1998931
Running Go benchmark...
go run main.go
Loop ops/sec: 4973217508
SHA-256 ops/sec: 2088447
Running Erlang benchmark...
erlc bench.erl
erl -noshell -s bench run -s init stop
Loop ops/sec: 1000000000
SHA-256 ops/sec: 1587301```

**Comment by harry-2401** [comment]
@04116 em thấy từ go 1.14 golang đã chuyển từ cooperative scheduling sang preemptive scheduling rồi ạ
ref: https://go.dev/doc/go1.14
e check thì với golang thì nếu goroutine chạy quá 10ms thì sẽ nhường chỗ cho các goroutine khác.
ref: https://github.com/golang/go/blob/go1.23.0/src/runtime/proc.go#L6187C1-L6188C1

em có thử một case thì goroutine không chặn các goroutine khác, nhưng nếu để 10ms thì có 1 goroutine bị block
`package main

import (
	"fmt"
	"runtime"
	"time"
)

func counter(i *int) {
	for j := range 100000000000000 {
		*i = j
	}
}

func main() {
	runtime.GOMAXPROCS(1)

	i := 0
	j := 0
	go counter(&i)
	go counter(&j)

	time.Sleep(20 * time.Millisecond)

	fmt.Printf("i = %d, j = %d\n", i, j)
}
`

**Reply by 04116** [reply]
chú chuẩn rồi, anh vừa check docs với thử lại 👍
